lib/utils.js

```diff
@@ -372,6 +372,47 @@ exports.merge = function merge (to, from) {
 var toString = Object.prototype.toString;
 
 /*!
+ * Applies toObject recursively.
+ *
+ * @param {Document|Array|Object} obj
+ * @return {Object}
+ * @api private
+ */
+
+exports.toObject = function toObject (obj) {
```

Comment:
@vkarpov15 

There are some reasons why I don't use `utils.clone()`.

#### 1. `utils.clone()` executes `mongoose.Types.Buffer.toObject()`

The following test case will be failed.

https://github.com/Automattic/mongoose/blob/924f79f1f42fa13bd1c417e52e6a0a8414a97f9b/test/document.populate.test.js#L399

Because queried by `mongoose.Types.Buffer`, and `mongoose.Types.Buffer.toObject()` converts like following.

```
{ _bsontype: 'Binary',
  sub_type: 0,
  position: 5,
  buffer: <Buffer 69 74 65 6d 31> }
```

#### 2. Behavior of `{ retainKeyOrder: 1 }` option is different between `utils.cloneObject()` and `Document.toObject()`

If you use `utils.clone()`, the following test case will be failed because a document cannot be updated by different orders between query and sub-document's properties.

https://github.com/Automattic/mongoose/blob/924f79f1f42fa13bd1c417e52e6a0a8414a97f9b/test/query.test.js#L1153

```
  1) Query querying/updating with model instance containing embedded docs should work (#454) works:

      Uncaught AssertionError: "hello" == "goodbye"
      + expected - actual

      +goodbye
      -hello

      at /home/naoina/work/src/github.com/Automattic/mongoose/test/query.test.js:1180:22
      at handleCallback (/home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/lib/utils.js:95:12)
      at /home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/lib/collection.js:861:5
      at handleCallback (/home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:234:5)
      at nextFunction (/home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:604:5)
      at /home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:538:7
      at queryCallback (/home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:215:5)
      at Callbacks.emit (/home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:94:3)
      at null.messageHandler (/home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:235:23)
      at Socket.<anonymous> (/home/naoina/work/src/github.com/Automattic/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:259:22)
      at Socket.EventEmitter.emit (events.js:95:17)
      at Socket.<anonymous> (_stream_readable.js:746:14)
      at Socket.EventEmitter.emit (events.js:92:17)
      at emitReadable_ (_stream_readable.js:408:10)
      at emitReadable (_stream_readable.js:404:5)
      at readableAddChunk (_stream_readable.js:165:9)
      at Socket.Readable.push (_stream_readable.js:127:10)
      at TCP.onread (net.js:528:21)
```

So, you will try to use `utils.clone(obj, { retainKeyOrder: 1 })` in order to keep the key orders of properties.

https://github.com/Automattic/mongoose/blob/924f79f1f42fa13bd1c417e52e6a0a8414a97f9b/lib/utils.js#L226

However, you will get the different failure message because a document cannot be got due to different orders between sub-document's properties and query.

```
  1) Query querying/updating with model instance containing embedded docs should work (#454) works:
     Uncaught TypeError: Cannot read property 'comments' of null
      at /home/naoina/work/src/github.com/Automattic/mongoose/test/query.test.js:1169:31
      at /home/naoina/work/src/github.com/Automattic/mongoose/lib/query.js:1172:16
      at /home/naoina/work/src/github.com/Automattic/mongoose/node_modules/kareem/index.js:109:16
      at process._tickCallback (node.js:415:13)
```

Also you can try following code to reproduce this behavior.

```js
var mongoose = require('mongoose');
var Sku = mongoose.model('Sku', {name: String});
var sku = new Sku({name: "item1"});
console.log(sku.toObject());
console.log(sku.toObject({ retainKeyOrder: 1 }));
process.exit(0);
```